### PR TITLE
Deadlines

### PR DESCRIPTION
**Pull request by frisi**
hi guido

i started working on collective.contentrule.mustread (which uses the schedule_must_read api of collective.mustread). i'll add a github repo for this soon so you can see how the api is used there.

in the meantime i'd love to get your feedback on this pr.
especially discuss the comments marked with `@guido` (as `TODO` did not survive code-analysis ;-)

### CODE REVIEW

**Comment by frisi** [comment]
@gyst tests pass, but the travis build seems to not terminate properly. could you please assist here, i don't know the ci setup well enough

**Comment by ale-rt** [comment]
Related #4

**Comment by ale-rt** [comment]
I rebased it on master after merging #4.

File: README.rst [hunk 1]
```diff
@@ -70,6 +70,21 @@ The database storage provides a rich API as specified in ``collective.mustread.i
 
 Note that the API design already contains specifications for some future extensions that have not been implemented yet.
```
Comment (gyst) [inline README.rst]: Is that still true?

File: src/collective/mustread/interfaces.py [hunk 2]
```diff
@@ -59,22 +59,22 @@ class ITracker(Interface):
     - mark_read(obj, 'maryjane')
     - has_read(obj, 'johndoe')
     - who_read(obj)
-
-    This is the initial implementation. It is not required (and at this stage
-    not even possible) to schedule a ``must_read(obj)`` to track reads.
+    - uids_read('johndoe')
 
     Scheduled for later extension:
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: I guess this is now in and the heading should be removed?

File: src/collective/mustread/interfaces.py [hunk 3]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
         If ``force_deadline`` is False, returns scheduled users who
         did not read the object at all.
 
+        @guido: i'd suggest to remove force_deadline as we have the get_stats
+        method to get a detailed report which users read and when
+
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: This was added as per your spec. I'm neutral, but if you want to remove `force_deadline` here IMO the `deadline` arg should be removed from `schedule_must_read` also. Either this application has a deadline concept and a way to query for that, or else the whole deadline thingy is handled by consuming code based on datetime reporting and this application itself does not care.

File: src/collective/mustread/interfaces.py [hunk 4]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+        If context is given, limit the result to objects located within
+        this context. If context is ``None`` search within the plonesite.
+
+        If userid is given, limit results to objects with an open read request
+        for this userid.
+
+        @guido: if we implement the force_deadline here, we'd have an easy way
+        to list all items with missed deadlines too. alternatively we can make
+        get_report more powerful/complex
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: If `deadline` is a meaningful concept in the business domain, as it appears to be, I'm in favor of adding it here as well. I think it's great to have well supported code paths to enforce read-before-deadline without having to write report parsers in consuming code.

File: src/collective/mustread/interfaces.py [hunk 5]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+
+        :param context: Context to search for objects with open read requests
+        :type context: Content object (must be IUUID resolvable)
+        :param userid: Userid to search for open read requests for.
+        :type userid: string
+        :returns: List of content objects with open read requests
+        :rtype: Generator
+        '''
+        # xxx guido: generator or list? (we intend to use all elements anyway)
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: Even if you use them all, if using means applying a filter and discarding 90% of entries, then a generator is more efficient.

File: src/collective/mustread/interfaces.py [hunk 6]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+                      type, path, ...}]``
+        :rtype: Generator
+        '''
+        # xxx guido: generator or list? (we intend to use all elements anyway)
+
+    def unschedule_must_read(obj=None, userids=None):
+        '''Maintenance method to remove open must-read requests for an object
+        and/or a specified userid
+
+        @guido decide if we need/want this
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: LOL you decide! I don't need `schedule_must_read` myself so I also don't need `unschedule_must_read`.

File: src/collective/mustread/interfaces.py [hunk 7]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+        and/or a specified userid with status 'not-read' and leaving the
+        read_at date empty.
+        this way they object is not listed as what_to_read
+        and the user ist not listed in who_must_read
+
+        this would allow admins to clear certain objects form the must-read
+        list or accept that a user (that might be ill) will not be required to
+        read a content item
+
+        @guido decide if we need/want this
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: Definitely not, since it duplicates `unschedule_must_read` and the approach suggested in this docstring is far more convoluted.

File: src/collective/mustread/models.py [hunk 8]
```diff
@@ -17,10 +17,12 @@ class MustRead(Base):
     userid = Column(Unicode(255), nullable=True)
     read_at = Column(DateTime, nullable=True)
     deadline = Column(DateTime, nullable=True)
+    scheduled_at = Column(DateTime, nullable=True)
+    scheduled_by = Column(String(255))
```
Comment (gyst) [inline src/collective/mustread/models.py:21]: This is very implicit, since it's not part of the interface contract of `schedule_must_read`. Which means you're implicitly adding the `request` to the interface contract of a backend component which is request-agnostic. In other words this column implies you're turning a database utility into a browser view. This will make testing and debugging much more difficult. If you need `scheduled_by`, add it to the contract of `schedule_must_read`.

File: src/collective/mustread/models.py [hunk 9]
```diff
@@ -17,10 +17,12 @@ class MustRead(Base):
...
     status = Column(String(255), nullable=True)
     uid = Column(String(255))
     type = Column(String(255), nullable=True)
-    title = Column(Unicode(255), nullable=True)
+    title = Column(Unicode(255), nullable=True)  # xxx remove?
     path = Column(String(255), nullable=True)
-    site_name = Column(Unicode(255), nullable=True)
-    info = Column(Unicode(255), nullable=True)
+    site_name = Column(Unicode(255), nullable=True)  # xxx remove?
+    info = Column(Unicode(255), nullable=True)  # xxx remove?
```
Comment (gyst) [inline src/collective/mustread/models.py]: Yes all of that cruft is copied over from the auditlog. Feel free to strip it down.

File: src/collective/mustread/interfaces.py [hunk 10]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+        method to get a detailed report which users read and when
+
         :param obj: Object that should be read by users
         :type obj: Content object (must be IUUID resolvable)
         :param force_deadline: Whether to ignore reads after the deadline
         :type force_deadline: Bool
-        :returns: Userids of users scheduled for reading which did not read
-        :rtype: List
+        :returns: Dictionary with userids of users scheduled for reading which
+                  did not read and their deadline (``{userid: deadline}``)
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: Here again you're working with the `deadline` concept, so I say keep `force_deadline` in the method signature for this method.

File: src/collective/mustread/interfaces.py [hunk 11]
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
...
+
+    def get_report(context=None, include_children=True, userid=None,
+                   start_date=None):
+        '''Return all mustread database entries for context.
+        (useful for exporting data or displaying detailed reports)
+
+        If context is None, defaults to plone site.
+
+        If include_children is ``True`` (which is the default), include
+        mustread records for child-objects of context.
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: Huh? The SQL layer has no containment concept. It does not know about Plone objects - those are squashed into UID strings and that's all the SQL ever needs to know about the content. Introducing this will have massive implementation implications, since you now cannot run the report in plain SQL anymore but will have to recurse into the Plone DB for any UID to determine what the child UIDs are, and then recurse again.

File: src/collective/mustread/tests/test_tracker.py [hunk 12]
```diff
@@ -195,12 +195,200 @@ def test_most_read_dates_verylongago_limit(self):
...
+        self.assertEqual(entry.scheduled_by, 'test_user_1_')
+        self.assertEqual(entry.deadline, self.deadline)
+        self.assertEqual(entry.scheduled_at.date(),
+                         datetime.date.today())
+
+    def test_schedule_must_read_existing(self):
+        """existing mustread request for an object do not get
+        overwritten, we return ids of users a request has
+        been written for
+        """
```
Comment (gyst) [inline src/collective/mustread/tests/test_tracker.py]: Smart, I overlooked that in the interface docstring but it makes sense.

File: src/collective/mustread/tracker.py [hunk 13]
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
...
+        current_user = api.user.get_current().id
+        new_users = []
+        for userid in userids:
+            if userid in existing_users:
+                # skip uses that already read the item
+                # or have an open mustread request
+                log.info('user {0} already has read / has to read {1}'.format(
+                    userid, path))
+                continue
+            # @guido: should we change the deadline for existing must_reads?
```
Comment (gyst) [inline src/collective/mustread/tracker.py]: Yes that makes sense. However then you should also update the `scheduled_at` (and possibly the `scheduled_by` but see above) and now you're painting yourself into a corner where it's possible that a document was read by a user before it was scheduled (because she read it on the first scheduling, before it got scheduled again). Of course that is already possible anyway: a user may have read an item before it got scheduled at all. Do you have test coverage for that edge case?

File: src/collective/mustread/tracker.py [hunk 14]
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
...
+    def schedule_must_read(self, obj, userids, deadline):
+        # get existing must read items for this object
+        qry = self._read(uid=utils.getUID(obj))
+        existing_users = [m.userid for m in qry.all()]
+
+        path = '/'.join(obj.getPhysicalPath())
+        uid = utils.getUID(obj)
+        hostname = utils.hostname()
+        now = datetime.utcnow()
+        current_user = api.user.get_current().id
```
Comment (gyst) [inline src/collective/mustread/tracker.py]: This will bite you once the client asks you to do a cron job 'on behalf of' the CEO. See my comments above. But maybe it's not that big a deal.

File: src/collective/mustread/tracker.py [hunk 15]
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
...
+
+        if start_date:
+            query = query.filter(or_(
+                MustRead.scheduled_at >= start_date.date(),
+                MustRead.read_at >= start_date.date()))
+
+        if userid:
+            query = query.filter(MustRead.userid == userid)
+        if include_children:
+            query = query.filter(MustRead.path.startswith(path))
```
Comment (gyst) [inline src/collective/mustread/tracker.py]: Ah!

File: src/collective/mustread/tracker.py [hunk 16]
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
...
+            # (see discussion at http://stackoverflow.com/q/1958219/810427)
+            yield dict([(k, v) for k, v in item.__dict__.iteritems()
+                        if not k.startswith('_')])
+
+    def get_stats_csv(self, csvfile, context=None, recursive=True):
+        """export mustread database entries for all objects within context
+        csvfile is a file-like object with a write method
+        (see csv.writer documentation)
+        xxx if guido does not like csv creation here in the api we can move
+        this to our contentrule
```
Comment (gyst) [inline src/collective/mustread/tracker.py]: Oh no this is cool. We're managing a SQL DB and this is a very valid view on that DB.

File: src/collective/mustread/interfaces.py (hunk 11 above)
Comment (gyst) [reply src/collective/mustread/interfaces.py]: Update: I now see the SQL layer is path aware and you're using that. Maybe that can be made more explicit in this interface contract by saying "child objects whose `path` startswith the context `path`"  or something like that. But I'm OK with this and withdraw my objection above. Just keeping it for reference.

**Review by gyst — CHANGES_REQUESTED** [review]
I haven't actually tested this, just reviewed the code, but looks fine to me in general, see more specific comments I made.

The main thing is to NOT force consumers to rely on the report generation for application logic. In other words: you use the report if you want to render a report, but you use the core API if you actually want to query the DB to support a specific execution path. IMO that implies supporting the `deadline` concept systematically throughout the CRUD API.

File: src/collective/mustread/tracker.py (hunk 16 above)
Comment (frisi) [inline src/collective/mustread/tracker.py]: thanks for your detailed feedback @gyst.
i'll iterate over your comments and will call you afterwards if any points are unclear to me

**Comment by frisi** [comment]
thanks for your detailed feedback @gyst.
i'll iterate over your comments and will call you afterwards if any points are unclear to me

File: README.rst (hunk 1 above)
Comment (frisi) [reply README.rst]: good catch - i'll rip that off

File: src/collective/mustread/interfaces.py (hunk 2 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: guess you meant `Scheduled for later extension:` . will remove that

File: src/collective/mustread/interfaces.py (hunk 5 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: taken... as what_to_read returns only a few objects (compared to the report) i'll make it return a list and stay with the generator for the report

File: src/collective/mustread/interfaces.py (hunk 11 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: added `(path starts with context's path)` to the docstring

File: src/collective/mustread/interfaces.py (hunk 6 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: i can imagine that when using the mustread functionality a lot you'll end up with quite a lot of objects returned by what_to_read although their deadline has passed a long time ago (think of employees that quit, where ill and "forget" about a reminder, etc).
we'll need a view (eg in collective.contentrules.mustread) that lists open requests and also allows privileged users to undo the schedulung of an open request or mark an open request as "accepted"

File: src/collective/mustread/interfaces.py (hunk 7 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: @gyst there is a semantic difference between: delete a read request from the database (unschedule_must_read - i.e. because it was scheduled by a mistake) and accepting that a user has not confirmed an open request (accept_not_read) 

however, i also agree, that adding another possible state `accepted` (or whatever we name it) adds extra complexity and is not required for my usecase either. so i'll drop that

File: src/collective/mustread/tracker.py (hunk 13 above)
Comment (frisi) [reply src/collective/mustread/tracker.py]: no. and i don't think it's necessary:
it could be that the document has been changed in the meantime and requires to be read again.
i'd say it's okay to require users to read an object after a read request has been sheduled, no matter if they marked it as read before.

File: src/collective/mustread/tracker.py (hunk 16 above)
Comment (frisi) [reply src/collective/mustread/tracker.py]: fine, i'll add it to the official api

File: src/collective/mustread/models.py (hunk 8 above)
Comment (frisi) [reply src/collective/mustread/models.py:21]: good point, i'll add scheduled_by to the signature and will not implicitly use current_user_id

File: src/collective/mustread/interfaces.py (hunk 6 above)
Comment (gyst) [reply src/collective/mustread/interfaces.py]: For that use case it makes more sense to introduce a `scheduled_after` or `deadline_after` arg to the relevant query methods.

File: src/collective/mustread/models.py (hunk 9 above)
Comment (frisi) [reply src/collective/mustread/models.py]: done.
title was selected in a query, but did not get accessed in the result set

File: src/collective/mustread/tracker.py (hunk 14 above)
Comment (frisi) [reply src/collective/mustread/tracker.py]: you are right, i changed scheduled_by to be explicitly passed as argument to schedule_must_read

**Comment by frisi** [comment]
i pushed some commits dealing with your review.
topics still open currently are:

* force_deadline
* whether to update an existing read request when scheduling another one or not.

i'll have a look at this after/while finishing https://github.com/collective/collective.contentrules.mustread

File: src/collective/mustread/interfaces.py (hunk 6 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: yes, a deadline_after=datetime would make sense for `what_to_read` indeed

File: src/collective/mustread/tracker.py (hunk 13 above)
Comment (gyst) [reply src/collective/mustread/tracker.py]: OK but then you'll have to also update the interface contract. Also it then does not make sense to return a user list anymore, since now this will axiomatically apply to all the userids you passed in.

File: src/collective/mustread/interfaces.py [hunk 17]
```diff
@@ -152,36 +150,40 @@ def most_read(days=None, limit=None):
...
         Calling this method and scheduling an object as 'must-read' enables
         tracking of 'unread' status for specific sets of users and querying
-        for those via the ``who_unread`` method.
+        for those via the ``who_must_read`` method.
 
         :param obj: Object that should be read by users
         :type obj: Content object (must be IUUID resolvable)
         :param userids: Userids of the users that should view the object.
-        :type userid: List
+        :type userids: List
```
Comment (gyst) [inline src/collective/mustread/interfaces.py:194]: This should be made optional, and then either default to the set of all users at schedule time, or `None` implying all users in the system at report time. Using schedule time is more robust - you don't want a new employee coming in and suddenly having a whole backlog of must_read items.

File: src/collective/mustread/interfaces.py [hunk 18]
```diff
@@ -152,36 +150,40 @@ def most_read(days=None, limit=None):
...
         :param deadline: Deadline before which users should view the object.
         :type deadline: datetime
+        :param by: Userid that scheduled the must read request
+        :type by: string
+        :returns: List of userids a mustread deadline has been scheduled for
+                  (existing deadlines do not get changed)
```
Comment (gyst) [inline src/collective/mustread/interfaces.py:200]: If you do decide to re-schedule this return value loses its value and should be dropped.

**Review by gyst — COMMENTED** [review]
See my newer comments on the fallout of updating existing scheduled must reads.

File: src/collective/mustread/interfaces.py (hunk 18 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py:200]: i guess it's better to not allow re-scheduling of existing read requests.
instead the frontend needs to delete existing mustreads and create new ones.

File: src/collective/mustread/interfaces.py [hunk 19]
```diff
@@ -59,22 +59,20 @@ class ITracker(Interface):
...
        track reads on that object, and query which users have failed to
        meet the requirment of reading the object in time:
 
-    - must_read(obj, ['johndoe', 'maryjane'], nextweek)
+    - schedule_must_read(obj, ['johndoe', 'maryjane'], nextweek)
     - mark_read(obj, 'johndoe')
-    - who_unread(obj)
+    - who_must_read(obj)
+    - what_to_read(context)
+    - what_to_read(userid)
```
Comment (gyst) [inline src/collective/mustread/interfaces.py]: Please also add `unschedule_must_read` to this docstring.

**Review by gyst — COMMENTED** [review]
Yes I think that is the most elegant and a good reason to have `unschedule_must_read`. However this implies that re-scheduling a must-read on a page that was already scheduled should throw an exception, and this needs to be documented in the interface contract and tested for.

But even when not rescheduling the return value should be removed, since again that is axiomatically identical to the method call argument.

File: src/collective/mustread/interfaces.py (hunk 19 above)
Comment (frisi) [reply src/collective/mustread/interfaces.py]: done in https://github.com/collective/collective.mustread/pull/1/commits/73f1a3dd808f4ea49ec70f8cce08319cd30e39a2

**Comment by frisi** [comment]
> Yes I think that is the most elegant and a good reason to have unschedule_must_read. However this implies that re-scheduling a must-read on a page that was already scheduled should throw an exception, and this needs to be documented in the interface contract and tested for.

i'd say we can keep it the way it is:

* an existing read request on page1 for user 'guido'
* call schedule_must_read(page1, ['guido', 'harald'], deadline)
* returns ['harald'], creates a deadline for harald (this must not necessarily be the same as the one of guido) and leaves the record for guido as it is.

this would allow a usecase where you grant additional users a role (or add them to a group) that matches the role-filter for a collective.contentrules.mustread action and re-fire the "request read confirmation event"
new users will get a mustread-db-record and receive a notification email. exisiting records for users will be unharmed.

> But even when not rescheduling the return value should be removed, since again that is axiomatically identical to the method call argument.

currently this is not the case. as we don't re-schedule existing records but skip them, the return value can be a subset of the userids passed to `schedule_must_read` in case there are existing mustread-records for.

actually i see that i'll need to change the current implementation so that records with state `read` are set to `must_read` and `read_at` gets cleared and scheduled_at, deadline, [and optionally scheduled_by] get set.

then we should be fine, shouldn't we?

**Comment by gyst** [comment]
> new users will get a mustread-db-record and receive a notification email. exisiting records for users will be unharmed.

Ack.

> as we don't re-schedule existing records but skip them, the return value can be a subset of the userids passed to schedule_must_read in case there are existing mustread-records for.

Ack.

> i'll need to change the current implementation so that records with state read are set to must_read and read_at gets cleared 

This I don't get. It contradicts the policy you've just outlined in the other statements I quoted. Or am I missing something?

**Comment by frisi** [comment]
> This I don't get. It contradicts the policy you've just outlined in the other statements I quoted. Or am I missing something?

currently we don't create a record with state mustread if there is an existing record for this object and user.
https://github.com/collective/collective.mustread/blob/deadlines/src/collective/mustread/tracker.py#L113

i'd change this to only skip the record if there is an existing read request (`scheduled_at != None`  no matter if it's marked as read or still mustread)

if the existing record is just a read-marker (`read_at != None`, `scheduled_at==None` then we create a must-read record for the person.

this way we omit the edge-case that no read-requests get sheduled for persons that read the object by accident (implementations such as yours in quaive mark objects as read unnoticed by the end-user) _before_ the read-request got created.

the reliable contract is:
* if a requests got scheduled before, it won't be changed
* if no request got scheduled before, we'll create one

**Comment by gyst** [comment]
I got that scenario, but I think it's confusing. If you schedule something as must-read sometime *after* you publish the page, why would that invalidate existing reads? Keeping the existing read record would be consistent with the scenario where we don't update the record if a user read a page between a first must-read schedule and a second must-read call on the same page.

In all of these scenarios, if the admin really wants to force re-reads they can call unschedule_must_read - oh no that's different again. You'd need a `clear_reads()` for that instead :-(.

OK I don't have an easy consistent solution for all these edge cases. And please don't feel blocked by my comments, I'm not making them forcefully but rather playfully since I enjoy doing it. Consider yourself the package maintainer and do it the way seems right to you...

**Comment by frisi** [comment]
thanks for your feedback and letting me decide about the (un)schedule api @gyst 

i tried to my best to provide a consitent and slim api. see the commit message for details: https://github.com/collective/collective.mustread/pull/1/commits/b6ca3bc2927e8e9b3ddf78096f0cca8b1edfafcb

i rebased the branch with master and merged my upgrade step with the one @ale-rt provided in the meantime to make it ready to merge.

could you please give me your feedback on the api changes done in the last commit.
if you're ok with that, please merge or let me know that i can do so.

thanks!

**Comment by frisi** [comment]
@gyst, i'd like to deploy this today. do you consider it ready to merge into master or do you have any objections?

**Comment by frisi** [comment]
thanks for merging
@gyst: could you please make a release?
or grant me pypi privileges? - user frisi

**Comment by gyst** [comment]
@frisi I've made you Owner on pypi
